### PR TITLE
Integrate HTTP Traffic uploader

### PR DESCRIPTION
**Pull request by DoomTaper**
## Description

+ This is an uploader which can upload http traffic from tool's report folders to owtf db, it has both CLI and GUI support to do so.
+ just after skipfish and hoppy finishes it uploads the data from the tool's output folder to owtf db
+ Supported tools: hoppy, skipfish, arachni, w3af, burpsuite
## Reviewers

@viyatb @DePierre 
## How Has This Been Tested?

I have supported tools' output report with me to test its functionality.
**CLI**:
./owtf.py --upload <tool_name> --path <tool's output path> <target_name>
**GUI**:
There is button "http uploader" on each target's page.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Other

### Checklist:

- [x] My code follows the code style (modified PEP8) of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

---
This change is [Reviewable](https://reviewable.io/reviews/owtf/owtf/835)

### CODE REVIEW

File: framework/interface/ui_handlers.py [hunk 1]
```diff
@@ -343,7 +343,21 @@ def get(self, extension=""):
 
 
 class PluginOutput(custom_handlers.UIRequestHandler):
-    SUPPORTED_METHODS = ['GET']
+    SUPPORTED_METHODS = ['GET', 'POST']
+
+    def post(self, target_id=None):
+        """ This function is a test function rght now to test the
```
Comment (DePierre) [inline framework/interface/ui_handlers.py]: A test function, how so? Should it be included in the OWTF codebase?

File: framework/interface/ui_handlers.py [hunk 2]
```diff
@@ -343,7 +343,21 @@ def get(self, extension=""):
...
+        http uploading functionality
+        """
+        tool_name = self.get_argument('toolname')
+        dir_path = self.get_argument('pathname')
+        try:
+            pUploader = PluginUploader(tool_name)
+            pUploader.init_uploader(dir_path)
+            pUploader.OWTFDBUpload(target_id=target_id)
+        except Exception as e:
```
Comment (DePierre) [inline framework/interface/ui_handlers.py]: Why is it needed to catch everything?

File: framework/lib/cli_options.py [hunk 3]
```diff
@@ -206,6 +206,17 @@ def parse_options(cli_options, valid_groups, valid_types):
         action='store_true',
         help='Run OWTF without its Web UI.')
     parser.add_argument('Targets', nargs='*', help='List of Targets')
+    parser.add_argument(
+        "--upload",
+        dest="ToolName",
```
Comment (DePierre) [inline framework/lib/cli_options.py]: Snake case.

File: framework/lib/cli_options.py [hunk 4]
```diff
@@ -206,6 +206,17 @@ def parse_options(cli_options, valid_groups, valid_types):
...
+        default=None,
+        help="The tool name of which you want to upload the report to OWTF Db "
+            "supported tools are (w3af, arachni, ...)")
+    parser.add_argument(
+        "--path",
+        dest="DirPath",
```
Comment (DePierre) [inline framework/lib/cli_options.py]: Snake Case.

File: owtf.py [hunk 5]
```diff
@@ -165,6 +165,13 @@ def process_options(user_args):
         if target[0] == "-":
             usage("Invalid Target: " + target)
 
+    if any([arg.ToolName, arg.DirPath]):
+        if not(all([arg.ToolName, arg.DirPath])):
+            usage("Provide both --upload as well as --path argument along with the target")
+        elif len(scope)!=1:
```
Comment (DePierre) [inline owtf.py]: Whitespace surrounding operators.

File: plugins/web/active/HTTP_Methods_and_XST@OWTF-CM-008.py [hunk 6]
```diff
@@ -17,4 +18,8 @@ def run(PluginInfo):
     resource = ServiceLocator.get_component("resource")
     Content += plugin_helper.CommandDump('Test Command', 'Output', resource.GetResources('ActiveHTTPMethods'),
                                          PluginInfo, Content)
+    plugin_output_dir = target.GetPath('plugin_output_dir')
+    pUploader = PluginUploader('hoppy')
```
Comment (DePierre) [inline plugins/web/active/HTTP_Methods_and_XST@OWTF-CM-008.py]: Snake Case

File: framework/db/plugin_uploader.py [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+
+	def __init__(self, tool_name):
+		self.supported_tools = ['w3af', 'skipfish', 'arachni', 'hoppy', 'burpsuite']
+		#self.full_parse_tools = ['w3af', 'skipfish', 'arachni']
+		self.db = ServiceLocator.get_component("db")
+		self.transaction = ServiceLocator.get_component("transaction")
+		self.re_extras = re.compile(r"(.*?) (\/.*?||\w+.*?) HT.*?")
+		if tool_name in self.supported_tools:
+			self.tool_name = tool_name
+			self.ptp = PTP(tool_name)
```
Comment (DePierre) [inline framework/db/plugin_uploader.py]: Please consider catching all PTP exception like in https://github.com/owtf/owtf/blob/develop/framework/plugin/plugin_handler.py#L279

If the Plugin Uploader is not mandatory, consider doing the same as in https://github.com/owtf/owtf/blob/develop/framework/plugin/plugin_handler.py#L281

File: framework/config/config.py [hunk 8]
```diff
@@ -127,7 +127,10 @@ def ProcessOptionsPhase1(self, options):
 
     def ProcessOptionsPhase2(self, options):
         target_urls = self.LoadTargets(options)
-        self.load_works(target_urls, options)
+        if(options['UploadTool']):
```
Comment (jack17529) [inline framework/config/config.py]: @DePierre does it needs a review or something needs to be changed?

Comment (viyatb) [inline framework/config/config.py]: @DoomTaper wherever possible, please change the functions and variable names to be `var_name` type and classes to be `MyClass` case.

File: framework/db/plugin_uploader.py [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+		else:
+			self.tool_name = None
+			raise Exception("%s tool not supported by this plugin" % tool_name)
+
+	def init_uploader(self, pathname):
+		parsed_data = self.ptp.parse(pathname)
+		self.parsed_data = parsed_data[-1]['transactions']
+
+	@target_required
+	def OWTFDBUpload(self, target_id=None):
```
Comment (viyatb) [inline framework/db/plugin_uploader.py]: Function naming.

File: framework/db/plugin_uploader.py [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+			Url = target_url + extras[1]
+			# Converting to latin1 to avoid ascii encoding errors
+			if self.tool_name == 'burpsuite': data['body'] = data['body'].encode('latin1', 'ignore')
+			transaction_model = models.Transaction(
+				url=Url,
+				scope=None,
+				method=Method,
+				data=None,
+				time=None,
+				time_human="Uploaded by " + self.tool_name, # to distinguish uploader reports in transaction table
```
Comment (viyatb) [inline framework/db/plugin_uploader.py]: Why are you storing "Uploading by tool" in the time_human field?

File: framework/interface/templates/target.html [hunk 11]
```diff
@@ -29,11 +29,39 @@ <h3 id="rankingTargetInfo"></h3>
...
+            Supported Tools <br><br>
+  <input type="radio" name="tool" value="skipfish" checked> Skipfish<br>
+  <input type="radio" name="tool" value="hoppy"> Hoppy<br>
+  <input type="radio" name="tool" value="w3af"> W3af<br>
+  <input type="radio" name="tool" value="arachni"> Arachni<br>
+  <input type="radio" name="tool" value="burpsuite"> Burpsuite
+</form>
+        </div>
+        <div class="modal-header">
+        <button id="httpUpload" class="btn btn-success" onclick="HTTPUpload();">Upload </button></div>
```
Comment (viyatb) [inline framework/interface/templates/target.html]: For JavaScript, naming should follow camelCase.

File: framework/interface/templates/target.html [hunk 12]
```diff
@@ -115,6 +143,30 @@ <h4 class="modal-title" id="pluginOutputFilterModalLabel">Advanced Filter</h4>
...
 
 <script>
+function HTTPUpload() {
+    var selectedVal = "";
+    var ReportPath = $("input[id='ReportPath']").val();
+    var selected = $("input[type='radio'][name='tool']:checked");
+    if (selected.length > 0) {
+        selectedVal = selected.val();
+    }
+    $('#uploaderModal').modal('hide')
```
Comment (viyatb) [inline framework/interface/templates/target.html]: Needs `;` after the statement.

File: framework/interface/templates/target.html [hunk 13]
```diff
@@ -115,6 +143,30 @@ <h4 class="modal-title" id="pluginOutputFilterModalLabel">Advanced Filter</h4>
...
+
+    $.ajax({
+    url: mySpace.poutput_ui_url+'?'+$.param(mySpace.filter_data, true),
```
Comment (viyatb) [inline framework/interface/templates/target.html]: Indentation.

**Comment by viyatb** [comment]
@DoomTaper can you update the PR?

File: framework/core.py [hunk 14]
```diff
@@ -216,6 +217,12 @@ def initialise_framework(self, options):
         self.start_proxy(options)  # Proxy mode is started in that function.
         # Set anonymised invoking command for error dump info.
         self.error_handler.SetCommand(OutputCleaner.anonymise_command(command))
+        # upload tool report
+        if options['UploadTool']:
+            pUploader = PluginUploader(options['UploadTool'])
+            pUploader.init_uploader(options['ReportUploadPath'])
```
Comment (DoomTaper) [inline framework/core.py]: It has been already taken care [Line 168](https://github.com/DoomTaper/owtf/blob/uploader/owtf.py#L168)

File: framework/db/plugin_uploader.py (hunk 10 above)
Comment (DoomTaper) [inline framework/db/plugin_uploader.py]: I have no preference. You tell me how to make it different when transactions are uploaded instead of proxying?

Comment (viyatb) [inline framework/db/plugin_uploader.py]: `time_human` just indicates human readable time. Store "NA" there.

**Comment by viyatb** [comment]
@DoomTaper LGTM. @DePierre ?

File: framework/core.py [hunk 15]
```diff
@@ -216,6 +217,12 @@ def initialise_framework(self, options):
         self.start_proxy(options)  # Proxy mode is started in that function.
         # Set anonymised invoking command for error dump info.
         self.error_handler.SetCommand(OutputCleaner.anonymise_command(command))
+        # upload tool report
+        if options['UploadTool']:
```
Comment (viyatb) [inline framework/core.py]: @DoomTaper

File: framework/db/plugin_uploader.py [hunk 16]
```diff
@@ -0,0 +1,143 @@
...
+		if(target_id is None):
+			print("Target id is None, aborting uploader!!")
+			return
+		if self.parsed_data is None:
+			print("Nothing found in the file, aborting uploader!!")
+			return
+		self.upload_checks()
+		target_url = ServiceLocator.get_component("target").GetTargetURLForID(target_id)
+		for data in self.parsed_data:
+			extras = self.re_extras.findall(data['request'])[0]
```
Comment (viyatb) [inline framework/db/plugin_uploader.py]: @DoomTaper

File: framework/db/plugin_uploader.py [hunk 17]
```diff
@@ -0,0 +1,143 @@
...
+			Method = extras[0]
+			Url = target_url + extras[1]
```
Comment (viyatb) [inline framework/db/plugin_uploader.py]: @DoomTaper

File: framework/db/plugin_uploader.py [hunk 18]
```diff
@@ -0,0 +1,143 @@
...
+			# Converting to latin1 to avoid ascii encoding errors
+			if self.tool_name == 'burpsuite': data['body'] = data['body'].encode('latin1', 'ignore')
```
Comment (viyatb) [inline framework/db/plugin_uploader.py]: @DoomTaper

File: framework/db/plugin_uploader.py [hunk 19]
```diff
@@ -0,0 +1,143 @@
...
+				pass
+		else:
+			print("Uploading data is approx '%0.1f%%' size of free space available" % percent_size)
+			cont = raw_input("Do you want to continue? y/n \n")
+			if cont == 'n':
+				raise Exception("Uploading aborted!")
+			else:
+				pass
+
+	def total_size(self, o, handlers={}, verbose=False):
```
Comment (viyatb) [inline framework/db/plugin_uploader.py]: @DoomTaper

File: framework/db/plugin_uploader.py [hunk 20]
```diff
@@ -0,0 +1,143 @@
...
+		def sizeof(o):
+			if id(o) in seen:       # do not double count the same object
+				return 0
+			seen.add(id(o))
+			s = getsizeof(o, default_size)
+			if verbose:
+				print(s, type(o), repr(o), file=stderr)
+			for typ, handler in all_handlers.items():
+				if isinstance(o, typ):
+					s += sum(map(sizeof, handler(o)))
```
Comment (viyatb) [inline framework/db/plugin_uploader.py]: @DoomTaper

File: framework/lib/cli_options.py (hunk 3 above)
Comment (viyatb) [inline framework/lib/cli_options.py]: @DoomTaper

File: framework/lib/cli_options.py (hunk 4 above)
Comment (viyatb) [inline framework/lib/cli_options.py]: @DoomTaper

File: plugins/web/active/HTTP_Methods_and_XST@OWTF-CM-008.py (hunk 6 above)
Comment (viyatb) [inline plugins/web/active/HTTP_Methods_and_XST@OWTF-CM-008.py]: @DoomTaper

**Comment by DePierre** [comment]
@DoomTaper As @viyatb commented, I had some questions regarding your PR that you didn't answer. It's been a while and I don't remember everything.

I am almost certain that some of my questions don't make any sense but I would appreciate if you could answer them anyway, even if it's to say "this doesn't make any sense". At least I would know that someone read them haha.

**Comment by DoomTaper** [comment]
I wont be able to continue this, hence closing PR

**Comment by viyatb** [comment]
@DoomTaper I can merge with additional changes from my side if you feel you are not able to work on this PR ;)

**Pull request closed without merging**